test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs

```diff
@@ -76,6 +77,144 @@ public void Save_NullSession_NonEmptyDictionary_Throws()
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(InvalidDictionaryTypes))]
+        public void EnsureObjectCanBeSerialized_InvalidDictionaryType_Throws(object value, Type type)
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+
+            // Act & Assert
+            var exception = Assert.Throws<InvalidOperationException>(() =>
+            {
+                testProvider.EnsureObjectCanBeSerialized(value);
+            });
+            Assert.Equal($"The dictionary with TKey {type} cannot be serialized to Session by '{typeof(SessionStateTempDataProvider).FullName}'.",
+                exception.Message);
+        }
+
+        public static TheoryData<object> ValidTypes
+        {
+            get
+            {
+                return new TheoryData<object>
+                {
+                    { 10 },
+                    { new int[]{ 10, 20 } },
+                    { "FooValue" },
+                    { new Uri("http://Foo") },
+                    { Guid.NewGuid() },
+                    { new List<string> { "foo", "bar" } },
+                    { new DateTimeOffset() },
+                    { 100.1m },
+                    { new Dictionary<string, int>() },
+                    { new Uri[] { new Uri("http://Foo"), new Uri("http://Bar") } }
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(ValidTypes))]
+        public void EnsureObjectCanBeSerialized_ValidType_DoesNotThrow(object value)
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+
+            // Act & Assert (Does not throw)
+            testProvider.EnsureObjectCanBeSerialized(value);
+        }
+
+        [Fact]
+        public void SaveAndLoad_WorksAsExpected()
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+            var inputGuid = Guid.NewGuid();
+            var inputDict = new Dictionary<string, string>
+            {
+                { "Hello", "World" },
+            };
+            var input = new Dictionary<string, object>
+            {
+                { "string", "value" },
+                { "int", 10 },
+                { "bool", false },
+                { "DateTime", new DateTime() },
+                { "Guid", inputGuid },
+                { "List`string", new List<string> { "one", "two" } },
+                { "Dictionary", inputDict },
+            };
+            var context = GetHttpContext(new TestSessionCollection(), true);
+
+            // Act
+            //System.Diagnostics.Debugger.Launch();
+            testProvider.SaveTempData(context, input);
+            var TempData = testProvider.LoadTempData(context);
+
+            // Assert
+            Assert.Equal("value", TempData["string"]);
+            Assert.Equal(10, Convert.ToInt32(TempData["int"]));
```

Comment:
Changed for everything except `int` and `List`.
The serializer assumes all `int` as `long` and returns `long` so directly casting doesn't work.
And for `List` directly casting works but `Assert.IsType` fails.
```C#
(IList<string>)TempData["List`string"] - works
Assert.IsType<IList<string>>(tempData["List`string"]) - fails